flake.nix

```diff
@@ -0,0 +1,56 @@
+{
+  description = "A Nix-flake-based Rust development environment";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/release-22.11";
+    flake-utils.url = "github:numtide/flake-utils";
+    rust-overlay.url = "github:oxalica/rust-overlay";
+  };
+
+  outputs =
+    { self
+    , nixpkgs
+    , flake-utils
+    , rust-overlay
+    }:
+
+    flake-utils.lib.eachDefaultSystem (system:
+    let
+      overlays = [
+        (import rust-overlay)
+        (self: super: {
+          rustToolchain =
+            let
+              rust = super.rust-bin;
+            in
+            if builtins.pathExists ./rust-toolchain.toml then
+              rust.fromRustupToolchainFile ./rust-toolchain.toml
+            else if builtins.pathExists ./rust-toolchain then
+              rust.fromRustupToolchainFile ./rust-toolchain
+            else
+              rust.stable.latest.default;
+        })
+      ];
+
+      pkgs = import nixpkgs { inherit system overlays; };
```

Comment:
we don't need overlays here, and nixpkgs is better accessed using `nixpkgs.legacyPackages.${system}`

https://zimbatm.com/notes/1000-instances-of-nixpkgs for reference